plugins/names.go

```diff
@@ -476,7 +477,63 @@ func (n *GoNames) GoTypeOfResponse(md *desc.MethodDescriptor) gopoet.TypeName {
 // upper-case. If the given symbol starts with an underscore, the underscore is
 // replaced with a capital "X".
 func CamelCase(s string) string {
-	return generator.CamelCase(s)
+	// NB(jh): This is forked from generator.CamelCase in the protobuf runtime.
+	// It is forked to avoid deprecation warnings. That entire package is now
+	// deprecated and even prints warnings in its init function(!!). But its
+	// replacement ("protogen" in the google.golang.org/protobuf module) does
+	// not have an analog for this function.
+	if s == "" {
+		return ""
+	}
+	t := make([]byte, 0, 32)
+	i := 0
+	if s[0] == '_' {
+		// Need a capital letter; drop the '_'.
+		t = append(t, 'X')
+		i++
+	}
+	// Invariant: if the next letter is lower case, it must be converted
+	// to upper case.
+	// That is, we process a word at a time, where words are marked by _ or
+	// upper case letter. Digits are treated as words.
+	for ; i < len(s); i++ {
+		c := s[i]
+		if c == '_' && i+1 < len(s) && isASCIILower(s[i+1]) {
+			continue // Skip the underscore in s.
+		}
+		if isASCIIDigit(c) {
+			t = append(t, c)
+			continue
+		}
+		// Assume we have a letter now - if not, it's a bogus identifier.
+		// The next word is a sequence of characters that must start upper case.
+		if isASCIILower(c) {
+			c ^= ' ' // Make it a capital letter.
+		}
+		t = append(t, c) // Guaranteed not lower case.
+		// Accept lower case sequence that follows.
+		for i+1 < len(s) && isASCIILower(s[i+1]) {
+			i++
+			t = append(t, s[i])
+		}
+	}
+	return string(t)
+}
+
+// camelCaseSlice is like CamelCase, but the argument is a slice of strings to
+// be joined with "_".
+func camelCaseSlice(elem []string) string {
```

Comment:
Should this be public so that plugin writers can use?

Reply:
I tried to put most everything a plugin writer actually needs, with more concrete/semantic method names, in the `Names` type.

Also, this is so simple, I maybe should have just inlined it.